### PR TITLE
Ensuing synchronization with Astarte

### PR DESCRIPTION
**Pull request by sorru94**

### CODE REVIEW

File: lib/astarte_device_sdk/device_connection.c [hunk 1]
```diff
@@ -370,101 +400,9 @@ static void state_machine_connected_run(astarte_device_handle_t device)
...
-
-    if (interface->ownership == ASTARTE_INTERFACE_OWNERSHIP_DEVICE) {
-        ares = astarte_device_tx_stream_individual(device, interface_name, path, individual, NULL);
-        ASTARTE_LOG_COND_ERR(ares != ASTARTE_RESULT_OK, "Failed sending cached property: %s",
-            astarte_result_to_name(ares));
-    }
-}
-
-static void send_purge_device_properties(astarte_device_handle_t device)
+static int send_purge_device_properties(astarte_device_handle_t device)
```
Comment (lucaato) [inline lib/astarte_device_sdk/device_connection.c:403]: Like the other case, why you decided to use an int as result type

File: lib/astarte_device_sdk/device_caching.c [hunk 2]
```diff
@@ -78,6 +82,83 @@ static astarte_result_t append_property_to_string(introspection_t *introspection
...
+        ASTARTE_LOG_INF("No previous synchronization with Astarte present.");
+        goto exit;
+    }
+    if (ares != ASTARTE_RESULT_OK) {
+        ASTARTE_LOG_ERR("Fetch error for cached introspection: %s.", astarte_result_to_name(ares));
+        goto exit;
+    }
+
+    char read_sync[MAX(sizeof(SYNCHRONIZATION_VALUE_TRUE), sizeof(SYNCHRONIZATION_VALUE_FALSE))]
+        = { 0 };
```
Comment (harlem88) [inline lib/astarte_device_sdk/device_caching.c:110]: Have you considered to use an int value or Enum? So you can compare without use a `strncmp` in the line below.

File: lib/astarte_device_sdk/device_connection.c [hunk 3]
```diff
@@ -205,7 +208,7 @@ astarte_result_t astarte_device_connection_poll(astarte_device_handle_t device)
  *         Static functions definitions         *
  ***********************************************/
 
-static void setup_subscriptions(astarte_device_handle_t device)
+static int setup_subscriptions(astarte_device_handle_t device)
```
[the remark this answers is not in the document]
Comment (sorru94) [reply lib/astarte_device_sdk/device_connection.c:211]: Sure, it does not really matter in this instance as there is a single possible reaction to those operations, independently from the error source.
However, I agree that it looks better as an Astarte result.

File: lib/astarte_device_sdk/device_caching.c (hunk 2 above)
Comment (sorru94) [reply lib/astarte_device_sdk/device_caching.c:110]: I think I'll make it a boolean. I would not expect it to create confusion regarding the meaning.

File: lib/astarte_device_sdk/device_caching.c [hunk 4]
```diff
@@ -78,6 +80,67 @@ static astarte_result_t append_property_to_string(introspection_t *introspection
...
+        ASTARTE_LOG_ERR("Fetch error for cached introspection: %s.", astarte_result_to_name(ares));
+        goto exit;
+    }
+
+    if (!read_sync) {
+        ASTARTE_LOG_INF("No previous synchronization with Astarte present.");
+        *sync = false;
+    } else {
+        *sync = true;
+    }
```
Comment (harlem88) [inline lib/astarte_device_sdk/device_caching.c:114]: What do you think to leave only the log in the if block?

```suggestion
    if (!read_sync) {
        ASTARTE_LOG_INF("No previous synchronization with Astarte present.");
    }
    *sync = read_sinc;
```

File: lib/astarte_device_sdk/device.c [hunk 5]
```diff
@@ -103,6 +103,16 @@ astarte_result_t astarte_device_new(astarte_device_config_t *cfg, astarte_device
...
+
+    bool synchronization_completed = false;
+#if defined(CONFIG_ASTARTE_DEVICE_SDK_PERMANENT_STORAGE)
+    ares = astarte_device_caching_synchronization_get(&synchronization_completed);
+    if ((ares != ASTARTE_RESULT_OK) && (ares != ASTARTE_RESULT_NOT_FOUND)) {
+        ASTARTE_LOG_ERR("Synchronization state getter failure %s.", astarte_result_to_name(ares));
+        goto failure;
+    }
+#endif
+    handle->synchronization_completed = synchronization_completed;
```
Comment (harlem88) [inline lib/astarte_device_sdk/device.c:115]: Use `synchronization_completed` directly from handle, without allocate different variable.
```suggestion
    handle->synchronization_completed = false;
#if defined(CONFIG_ASTARTE_DEVICE_SDK_PERMANENT_STORAGE)
    ares = astarte_device_caching_synchronization_get(&handle->synchronization_completed);
    if ((ares != ASTARTE_RESULT_OK) && (ares != ASTARTE_RESULT_NOT_FOUND)) {
        ASTARTE_LOG_ERR("Synchronization state getter failure %s.", astarte_result_to_name(ares));
        goto failure;
    }
#endif
```

**Review by harlem88 — CHANGES_REQUESTED** [review]

**Review by harlem88 — APPROVED** [review]

**Review by lucaato — APPROVED** [review]

**Pull request merged**